### ISSUE TITLE
[Feature]: In-place recovery from backup

### ISSUE THREAD

**Issue by fullykubed** [cloudnative-pg/cloudnative-pg#5203]
### Is there an existing issue already for this feature request/idea?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new feature request to be evaluated.

### What problem is this feature going to solve? Why should it be added?

Currently,

> In CloudNativePG, you can't perform recovery in place on an existing cluster. Recovery is instead a way to bootstrap a new Postgres cluster starting from an available physical backup.

While it is amazing to have this capability, the ergonomics are not great, especially if needing to recover a database in an emergency scenario.

The current recovery mechanism has the following friction points:

1. Requires changing the cluster spec (`bootstrap.recovery` + `externalClusters`)

2. Requires creating a new cluster

3. Requires deleting the old cluster

### Describe the solution you'd like

It seems like a cluster that has the `backup` spec configured has all of the information it needs find available backups and initiate an in-place recovery.

To make it easier to recover an existing cluster to one of its backups, I'd propose the following:

1. Add a `kubectl-cnpg recover` command to the cnpg plugin that exposes the same fields as `bootstrap.recovery` on the `Cluster` spec.

2. When calling that command:

   1. Internally, reset the cluster by removing the pods and state.
   
   2. Execute the logic that is currently run when bootstrapping a new cluster with the `bootstrap.recovery` field set using the options supplied to the CLI.
   
For the end user, this would make recovery in an emergency situation much faster and less error-prone. In effect, it would require a user to just run a single command such as `kubectl cnpg recover <cluster> --source=myBackup --targetTime="2023-08-11 11:14:21.00000+02"`.

### Describe alternatives you've considered

Alternatively, you could supply a `recovery` field on the Cluster spec outside of the `bootstrap`. When that field changes, an in-place recovery is initiated.

### Backport?

No

### Are you willing to actively contribute to this feature?

No

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by xieydd** [comment]
Great and very useful features. But it is very important to confirm the correct restore to your target time.

**Comment by fullykubed** [comment]
Absolutely, I would imagine if implemented in the CLI, there would be a confirmation prompt that required re-entering / confirming the information.

I would also propose that the existing backups should _not_ be dropped on the restore.

I believe this is already the default behavior:

> The operator includes a safety check to ensure a cluster doesn't overwrite a storage bucket that contained information. A cluster that would overwrite existing storage remains in the state Setting up primary with pods in an error state. The pod logs show: ERROR: WAL archive check failed for server
recoveredCluster: Expected empty archive.

That way if you were to restore to the wrong backup / point-in-time, you could always retry without data loss.

Of course, this would require manual intervention in order to have backups working again after the restore, but I think this is fine given the primary goal here is simply to minimize the friction of executing the recovery itself.

**Comment by PrivatePuffin** [comment]
I agree the current design is simply not infrastructure-as-code compliant enough

**Comment by fullykubed** [comment]
@onedr0p Not sure what you mean by this:

> A way to have cnpg bootstrap a cluster from backups would be amazing. 

This feature already exists. This feature request is a way to initiate a recovery on a cluster that has already been bootstrapped.

**Comment by onedr0p** [comment]
> @onedr0p Not sure what you mean by this:
> 
> > A way to have cnpg bootstrap a cluster from backups would be amazing.
> 
> The ability to bootstrap a cluster from backups already exists. This feature request is a way to initiate a recovery on a cluster that has already been bootstrapped. That is currently not possible using native CNPG functionality.

Let me describe the scenario I run into...

My Kubernetes cluster managed by [Flux](https://fluxcd.io/) is all happy, pg cluster running great, backups are running. Now there is a disaster and I need to re-provision my Kubernetes cluster. After I have the new Kubernetes cluster ready to go and before I tell Flux to put everything back I need to change the CNPG Cluster manifest as described in my comments in the manifest I posted above. Ideally I would like CNPG to always try to recover on first bootstrap if there's a recovery found. It's the manually changing of the CNPG Cluster manifest that is hard to fit into a GitOps pattern.

If this isn't the same as what you are asking sorry but it did look related.

**Comment by onedr0p** [comment]
My request looks more like https://github.com/cloudnative-pg/cloudnative-pg/issues/5778 so I'll bring it over there, sorry for the noise!

**Comment by fullykubed** [comment]
@gbartolini Wanted to touch base on this one. This has quickly become one of the most 👍🏻 feature requests for the project, but it is still in the triage phase. I don't feel comfortable proceeding with an implementation until a project maintainer provides some directional guidance on the preferred approach here.

**Comment by nurbek-workfindme** [comment]
I like how it is done in [Percona Operator for MySQL](https://docs.percona.com/percona-operator-for-mysql/pxc/backups-restore.html#restore-the-cluster-with-point-in-time-recovery). You just define `Restore` custom resource with all required parameters, perform `kubectl apply` and recovery process is running. No need to even touch `Cluster` CR and IMO this is more declarative than running CLI command. As a bonus we can have history of recovery attempts as Kubernetes objects and in general it can be better integrated with monitoring tools. However, having both methods (CLI and Custom Resource) would be great!

**Comment by PrivatePuffin** [comment]
> I like how it is done in [Percona Operator for MySQL](https://docs.percona.com/percona-operator-for-mysql/pxc/backups-restore.html#restore-the-cluster-with-point-in-time-recovery). You just define `Restore` custom resource with all required parameters, perform `kubectl apply` and recovery process is running. No need to even touch `Cluster` CR and IMO this is more declarative than running CLI command. As a bonus we can have history of recovery attempts as Kubernetes objects and in general it can be better integrated with monitoring tools. However, having both methods (CLI and Custom Resource) would be great!

I'm not sure that works any better with IaC.

**Comment by fullykubed** [comment]
Having a Restore CRD wouldn't influence the ability of this to work IaC. In fact, it actually provides more options.

The `kubectl-cnpg recover` CLI command would simply create the CR. Additionally, you could now also create a Recovery using your existing favorite IaC tooling (`kubectl`, `helm`, `terraform`, etc.).

This also matches that pattern implemented in other popular backup / restore solutions for Kubernetes such as Velero.

**Comment by zifeo** [comment]
@dosubot still a relevant feature request.

**Comment by teeg82** [comment]
Definitely still relevant. I was pretty surprised this didn't exist already. This would greatly speed up mean time to recovery.

**Comment by AndrinGautschi** [comment]
For now, at least as long as this feature is not implemented/ready, one can recover "in-place" by changing the path in the backup bucket by attaching a timestamp (or some other addition) to it. 

```yaml
  backup:
    barmanObjectStore:
      destinationPath: "s3://bucket/cnpg/currenttimestamp/"
```

With this you can restart the cluster with the same name and have it recover from a previous state. Obviously this approach needs either someone manually changing this value in a disaster scenario, or having some kind of deploy script that does it automatically. 

```yaml
  bootstrap:
    recovery:
      source: clusterBackup
      database: dbname
      owner: dbowner
      secret:
        name: db-cluster-user-secret

  externalClusters:
    - name: clusterBackup
      barmanObjectStore:
        destinationPath: "s3://bucket/cnpg/previoustimestamp"
        endpointURL: "https://somestorage.com"
        serverName: postgres-cluster #can stay the same and is equal to the now recovering cluster name
        wal:
          maxParallel: 8
        s3Credentials:
          accessKeyId:
            name: postgres-backup-bucket-secret
            key: AWS_ACCESS_KEY_ID
          secretAccessKey:
            name: postgres-backup-bucket-secret
            key: AWS_SECRET_ACCESS_KEY
```

Obviously, this is still a dirty workaround and no proper replacement for a out of the box in-place recovery solution.

**Comment by fullykubed** [comment]
Not stale

**Comment by R-Studio** [comment]
Any news?

**Comment by 0xA1F1E** [comment]
Just came across this after googling for the same issue. I'm surprised that this feature request is more than a year old and has not been picked up yet, despite the demand for this feature 😅 This would be so useful for IaC deployments!

**Comment by ChrislerOne** [comment]
I also want to make sure that this would be a very helpful feature. Is there any chance that this is gonna be implemented soonish? Or at least that some of the more experience maintainers give some implementation advice that this can be picked up by another person.

**Comment by LokoMoloko98** [comment]
Please, this would be a great feature to have in the next release 🙏

**Comment by lieberlois** [comment]
Totally agree with the others here. Definitely not stale, personally I think the Restore CRD would be the way to go. Let it start a new Postgres timeline and we're good! 

Would be the preferred approach for GitOps.

**EDIT**: Thinking about this a bit more I also like the suggested `Alternatively, you could supply a recovery field on the Cluster spec outside of the bootstrap. When that field changes, an in-place recovery is initiated.`. Right now (at least for us) the tedious thing is having to distinguish between initializing and recovering at template time. For me it seems more convenient to mirror CPGO's behavior. First check if a backup is there, if yes, recover the latest (if nothing else is specified). If not, initialize a new cluster.

This `bootstrap` approach that only runs once on cluster creation feels a bit limiting to me. @gbartolini what do you think, would this be very hard to implement? If I find the time I'd be happy to assist, although I'd need some pointers, would be my first time contributing here.

**Comment by fullykubed** [comment]
@gbartolini Is there anything else you + team would need to get this out of the triage phase?

**Comment by knightly-flo** [comment]
+1 to this. Solution proposed by AndrinGautschi works, but it would be great to have a proper feature instead of this hack. Previously I've used CrunchyData's Postgres Operator, which has native support for both in-place and bootstrap methods of recovery, and it was quite useful. Also AFAIK other major Postgres Kubernetes operators support in-place recovery, so this isn't some kind of a novelty feature.

**Comment by Archisman-Mridha** [comment]
+1 to having this feature.

**Comment by andersen-k8s** [comment]
This feature is highly requested.

**Comment by LokoMoloko98** [comment]
@tharanathkartoza

**Comment by linwalth** [comment]
This would be a most useful feature to have.

**Comment by jtackaberry** [comment]
Not stale.

**Comment by madhavpcm** [comment]
This seems to be a very useful feature, however
> Internally, reset the cluster by removing the pods and state.

Wouldn't this delete the existing PVCs?

**Comment by lucian-student** [comment]
Not stale

**Comment by linwalth** [comment]
Still not stale and in my opinion still pertaining to #5778 :)

**Comment by okuuva** [comment]
Not stale

**Comment by lieberlois** [comment]
@gbartolini myself and many other people would love your opinion on this. I myself would be happy to contribute, but the inactivity in such an important issue doesn't give me the feeling that putting in the work is worth the effort.

We all understand maintaining this project is a bit of work, but maybe we can align on this? Happy to talk in person if you prefer that 🙂 cheers!
